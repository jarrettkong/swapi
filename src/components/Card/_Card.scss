@import '../../util/colors';

.Card {
	padding: 10px;
	background-color: white;
	color: black;
	text-align: left;
	border-radius: 2px;
	box-shadow: 0 0 5px black;

	& hr {
		border-style: none;
		border-bottom: 1px solid $main-color;
	}
}

.Card-header {
	position: relative;
}

.Card-favorite-btn {
  position: absolute;
  top: 3px;
  font-size: 20px;
  right: 0;
	background-color: transparent;
  border: 0;
  cursor:pointer;
  
  &:hover {
    color: $favorite-color;
  }
}

.Card-data {
	display: flex;
	justify-content: space-between;
	align-items: flex-start;
	margin: 15px 0;
}

.Card-title {
	margin: 0;
	font-size: 25px;
}

.Card-prop {
	font-weight: 600;
	margin-right: 50px;
	color: $empire-blue;
}

.Card-value {
	font-weight: 300;
	text-align: right;
	color: $light-grey;
}
