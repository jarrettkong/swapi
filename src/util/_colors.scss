$empire-blue: #445266;
$favorite-color: #d1000a;
$star-wars-yellow: #fed70f;
$main-color: #262626;
$background-color: #171315;
$light-grey: #7d7c77;
