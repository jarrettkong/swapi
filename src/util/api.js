export const fetchHomeworld = people => {
	const promises = people.map(person => {
		return fetch(person.homeworld)
			.then(res => res.json())
			.then(homeworld => ({ ...person, homeworld }));
	});
	return Promise.all(promises);
};

export const fetchResidents = planets => {
	const promises = planets.map(planet => {
		return fetchResidentNames(planet.residents).then(residents => ({ ...planet, residents }));
	});
	Promise.all(promises).then(planets => console.log(planets));
};

const fetchResidentNames = residents => {
	const promises = residents.map(resident => {
		return fetch(resident)
			.then(res => res.json())
			.then(data => data.name);
	});
	return Promise.all(promises);
};

export const fetchSpecies = people => {
	const promises = people.map(person => {
		return fetch(person.species[0])
			.then(res => res.json())
			.then(species => ({ ...person, species }));
	});
	return Promise.all(promises);
};

export const searchApi = term => {
	return fetch(`https://swapi.co/api/${term}`).then(response => {
		if (!response.ok) {
			throw new Error('Error adding grocery');
		}
		return response.json();
	});
};
