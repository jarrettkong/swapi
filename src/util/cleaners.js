export const cleanPeople = people => {
	return people.map(person => {
		return {
			Name: person.name,
			Species: person.species.name,
			Homeworld: person.homeworld.name,
			'Population of Homeworld': person.homeworld.population,
			favorite: false
		};
	});
};

export const cleanPlanets = planets => {
	return planets.map(planet => {
		return {
			Name: planet.name,
			Population: planet.population,
			Climate: planet.climate,
			Terrain: planet.terrain,
			Residents: planet.residents.join(', ') || 'N/A',
			favorite: false
		};
	});
};

export const cleanVehicles = vehicles => {
	return vehicles.map(vehicle => {
		return {
			Name: vehicle.name,
			Model: vehicle.model,
			Class: vehicle.vehicle_class,
			Passengers: vehicle.passengers,
			favorite: false
		};
	});
};
