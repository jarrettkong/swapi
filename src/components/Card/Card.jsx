import React from 'react';
import PropTypes from 'prop-types';
import './_Card.scss';

const Card = props => {
	const keys = Object.keys(props.info);
	const info = keys.slice(1, keys.length - 1);
	const name = props.info['Name'];

	return (
		<article className="Card">
			<div className="Card-header">
				<h3 className="Card-title">{name}</h3>
				<button className="Card-favorite-btn" onClick={() => props.toggleFavorite(name)}>
					<i className={`fab fa-rebel ${props.info.favorite ? 'favorited' : null}`} />
				</button>
			</div>
			<hr />
			<div>
				{info.map((infoKey, i) => (
					<p key={i} className="Card-data">
						<span className="Card-prop">{`${infoKey}`}</span>
						<span className="Card-value">{props.info[infoKey]}</span>
					</p>
				))}
			</div>
		</article>
	);
};

Card.propTypes = {
	info: PropTypes.object,
	toggleFavorite: PropTypes.func
};

export default Card;
