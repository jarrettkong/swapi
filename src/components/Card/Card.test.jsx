import React from 'react';
import Card from './Card';
import { mockPeople, mockPlanets, mockVehicles } from '../../util/mockData';
import { cleanPeople, cleanPlanets, cleanVehicles } from '../../util/cleaners';
import { shallow } from 'enzyme';

describe('Card', () => {
	it('should match the snapshot with people', () => {
		const mockInfo = cleanPeople(mockPeople);
		const wrapper = shallow(<Card info={mockInfo} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should match the snapshot with planets', () => {
		const mockInfo = cleanPlanets(mockPlanets);
		const wrapper = shallow(<Card info={mockInfo} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should match the snapshot with vehicles', () => {
		const mockInfo = cleanVehicles(mockVehicles);
		const wrapper = shallow(<Card info={mockInfo} />);
		expect(wrapper).toMatchSnapshot();
	});
});
