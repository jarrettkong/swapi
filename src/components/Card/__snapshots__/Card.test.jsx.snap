// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot with people 1`] = `ShallowWrapper {}`;

exports[`Card should match the snapshot with planets 1`] = `ShallowWrapper {}`;

exports[`Card should match the snapshot with vehicles 1`] = `ShallowWrapper {}`;
